pkg/image/external/external.go

```diff
@@ -0,0 +1,160 @@
+package external
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/rancher/rancher/pkg/controllers/management/k3sbasedupgrade"
+	"github.com/sirupsen/logrus"
+)
+
+type Source string
+
+const (
+	K3S  Source = "k3s"
+	RKE2 Source = "rke2"
+)
+
+var (
+	RKE2MinimumKubernetesVersion = &semver.Version{
+		Major: 1,
+		Minor: 21,
+		Patch: 0,
+	}
+)
+
+func GetExternalImages(rancherVersion string, externalData map[string]interface{}, source Source, os, arch string) ([]string, error) {
+	if source != K3S && source != RKE2 {
+		return nil, fmt.Errorf("invalid source provided: %s", source)
+	}
+
+	logrus.Infof("generating %s image list...", source)
+	externalImagesMap := make(map[string]bool)
+	releases, _ := externalData["releases"].([]interface{})
+	var compatibleReleases []string
+
+	for _, release := range releases {
+		releaseMap, _ := release.(map[string]interface{})
+		version, _ := releaseMap["version"].(string)
+		if version == "" {
+			continue
+		}
+
+		// Check that RKE2 uses at least Kubernetes v1.21+.
+		if source == RKE2 {
+			versionSemVer, err := semver.NewVersion(strings.TrimPrefix(version, "v"))
+			if err != nil {
+				continue
+			}
+			if versionSemVer.LessThan(*RKE2MinimumKubernetesVersion) {
+				continue
+			}
+		}
+
+		if rancherVersion != "dev" {
+			maxVersion, _ := releaseMap["maxChannelServerVersion"].(string)
+			maxVersion = strings.TrimPrefix(maxVersion, "v")
+			if maxVersion == "" {
+				continue
+			}
+			minVersion, _ := releaseMap["minChannelServerVersion"].(string)
+			minVersion = strings.Trim(minVersion, "v")
+			if minVersion == "" {
+				continue
+			}
+
+			// "IsNewerVersion" works for both k3s and rke2.
+			versionGTMin, err := k3sbasedupgrade.IsNewerVersion(minVersion, rancherVersion)
+			if err != nil {
+				continue
+			}
+			if rancherVersion != minVersion && !versionGTMin {
+				// Rancher version not equal to or greater than minimum supported rancher version.
+				continue
+			}
+
+			// "IsNewerVersion" works for both k3s and rke2.
```

Comment:
This comment is fine but that is why the package name was changed to include "k3sbased" since rke2 is k3s based and therefore it applies to both in the name.

Reply:
I thought the name was strange. Will remove.
![facepalm picard](https://user-images.githubusercontent.com/39320683/119906980-50790f80-bf1d-11eb-9970-5007abfd9998.png)